src/main/java/com/googleintern/wfm/ruleengine/action/RuleValidation.java

```diff
@@ -0,0 +1,286 @@
+  /** Group Truth of UserPoolAssignments data */
+  public static ImmutableList<UserPoolAssignmentModel> expectedUserPoolAssignments;
+
+  /** Constant string variables used for result outputs. */
+  private static final String RULE_COVERAGE_PERCENT_HEADER = "Coverage % for Rule Set:";
+
+  private static final String[] POOL_ASSIGNMENT_HEADER =
+      new String[] {"Case Pool ID", "Permission Set ID"};
+  private static final String[] WRONG_USER_POOL_ASSIGNMENT_HEADER =
+      new String[] {
+        "User Id",
+        "Workforce Id",
+        "Workgroup Id",
+        "Role Id",
+        "Skill Id",
+        "Expected Pool Assignments",
+        "Actual Pool Assignments"
+      };
+  private static final String PERCENTAGE_SIGN = "%";
+  private static final String CASE_POOL_ID_PREFIX = "{\"case_pool_id\":\"";
+  private static final String PERMISSION_SET_ID_PREFIX = "\",\"permission_set_id\":\"";
+
+  public RuleValidation(
+      Set<RuleModel> rules, List<UserPoolAssignmentModel> expectedUserPoolAssignments) {
+    this.rules = (ImmutableSet<RuleModel>) rules;
+    this.expectedUserPoolAssignments =
+        (ImmutableList<UserPoolAssignmentModel>) expectedUserPoolAssignments;
+    this.actualAssignedPermissionsByUserPoolAssignment =
+        assignPermissionsBasedOnGeneratedRules(expectedUserPoolAssignments);
+  }
+
+  public static void writeRuleValidationResultsIntoCsvFile(String outputCsvFilePath)
+      throws IOException {
+    File outputFile = new File(outputCsvFilePath);
+    Files.deleteIfExists(outputFile.toPath());
+    outputFile.createNewFile();
+
+    FileWriter outputFileWriter = new FileWriter(outputFile);
+    CSVWriter csvWriter = new CSVWriter(outputFileWriter);
+
+    ImmutableList<String[]> validationResult = convertRuleValidationResultsToString();
+
+    csvWriter.writeAll(validationResult);
+    csvWriter.close();
+  }
+
+  public static double calculateRulesCoverage(
+      ImmutableSet<UserPoolAssignmentModel> usersWithWrongAssignedPermissions) {
+    double numberOfWrongAssignment = usersWithWrongAssignedPermissions.size();
+    double totalNumberOfUsers = expectedUserPoolAssignments.size();
+    return (totalNumberOfUsers - numberOfWrongAssignment) / totalNumberOfUsers;
+  }
+
+  public static ImmutableSet<UserPoolAssignmentModel> findUncoveredUsers() {
+    ImmutableSet.Builder<UserPoolAssignmentModel> usersWithWrongAssignedPermissionsBuilder =
+        ImmutableSet.builder();
+    for (UserPoolAssignmentModel user : expectedUserPoolAssignments) {
+      if (!actualAssignedPermissionsByUserPoolAssignment.containsKey(user)
+          && user.poolAssignments().size() != 0) {
+        usersWithWrongAssignedPermissionsBuilder.add(user);
+      } else if (!user.poolAssignments()
+          .equals(actualAssignedPermissionsByUserPoolAssignment.get(user))) {
+        usersWithWrongAssignedPermissionsBuilder.add(user);
+      }
+    }
+    return usersWithWrongAssignedPermissionsBuilder.build();
```

Comment:
The whole for loop can be represented using lambda function: 

`expectedUserPoolAssignments.stream().filter(/**if statements*/).collect(toImmutableSet)`

Same elsewhere.